File: public/lib/generator/riffs.js
/**
 * riffs.js — Drum and melody riff variant generators.
 * Port of petri-note/internal/generator/riffs.go
 */

import { euclidean } from './euclidean.js';
import { markovMelody } from './markov.js';
import { createRng } from './core.js';

/**
 * drumRiff generates a drum pattern variant.
 * variant "A" = base pattern, "B" = more energy, "C" = less energy.
 * Drums keep the same hits/rotation across all variants for musical consistency.
 * If style is non-null, it overrides the hardcoded A/B/C adjustments.
 *
 * @param {string} variant - "A", "B", or "C"
 * @param {number} hits - Number of hits (K in Euclidean)
 * @param {number} steps - Total steps (N in Euclidean)
 * @param {number} rotation - Pattern rotation
 * @param {number} note - MIDI note number
 * @param {Object} params - Generator params (channel, velocity, duration, accent, etc.)
 * @param {Object} [style] - DrumVariant { hitsAdd, hitsMul, rotationAdd }
 * @returns {import('../pflow.js').NetBundle}
 */
export function drumRiff(variant, hits, steps, rotation, note, params, style) {
    // Copy params so we don't mutate the caller's object
    params = Object.assign({}, params);

    if (style && variant !== 'A') {
        hits = hits + (style.hitsAdd || 0);
        if (style.hitsMul && style.hitsMul !== 1.0) {
            hits = Math.floor(hits * style.hitsMul);
        }
        rotation = (rotation + (style.rotationAdd || 0)) % steps;
        if (hits < 1) hits = 1;
        if (hits > steps) hits = steps;
    } else {
        // Same pattern, vary velocity/duration for feel
        switch (variant) {
        case 'B':
            params.velocity = Math.min(127, params.velocity + 15);
            params.duration = Math.floor(params.duration * 1.4);
            break;
        case 'C':
            params.velocity = Math.max(40, params.velocity - 20);
            params.duration = Math.floor(params.duration * 0.6);
            break;
        }
    }

    const result = euclidean(hits, steps, rotation, note, params);
    return result.bundle;
}

/**
 * melodyRiff generates a melody pattern variant.
 * variant "A" = base, "B" = higher density, "C" = sparse/longer notes.
 *
 * @param {string} variant - "A", "B", or "C"
 * @param {Object} params - Generator params (scale, rootNote, channel, velocity, duration, density, seed, steps, chords, durationVariation, syncopation)
 * @returns {import('../pflow.js').NetBundle}
 */
export function melodyRiff(variant, params) {
    // Copy params so we don't mutate the caller's object
    params = Object.assign({}, params);

    const rng = createRng(params.seed);

    switch (variant) {
    case 'B':
        // Higher density, slight seed variation
        params.density = params.density * 1.4;
        if (params.density > 1.0) params.density = 1.0;
        params.seed = params.seed + 1000 + Math.floor(rng.float64() * 1000);
        break;
    case 'C':
        // Sparse, longer notes
        params.density = params.density * 0.5;
        params.duration = Math.floor(params.duration * 1.5);
        params.seed = params.seed + 2000 + Math.floor(rng.float64() * 1000);
        break;
    }

    const result = markovMelody(params);
    return result.bundle;
}

File: internal/generator/riffs.go
package generator

import (
	"math/rand"

	"beats-bitwrap-io/internal/pflow"
)

// DrumRiff generates a drum pattern variant.
// variant "A" = base pattern, "B" = same pattern with more energy (velocity/duration),
// "C" = same pattern, quieter/shorter.
// Drums keep the same hits/rotation across all variants for musical consistency.
// If style is non-nil, it overrides the hardcoded A/B/C adjustments.
func DrumRiff(variant string, hits, steps, rotation, note int, params Params, style *DrumVariant) *pflow.NetBundle {
	if style != nil && variant != "A" {
		hits = hits + style.HitsAdd
		if style.HitsMul != 0 && style.HitsMul != 1.0 {
			hits = int(float64(hits) * style.HitsMul)
		}
		rotation = (rotation + style.RotationAdd) % steps
		if hits < 1 {
			hits = 1
		}
		if hits > steps {
			hits = steps
		}
	} else {
		// Same pattern, vary velocity/duration for feel
		switch variant {
		case "B":
			params.Velocity = min(127, params.Velocity+15)
			params.Duration = int(float64(params.Duration) * 1.4)
		case "C":
			params.Velocity = max(40, params.Velocity-20)
			params.Duration = int(float64(params.Duration) * 0.6)
		}
	}

	result := Euclidean(hits, steps, rotation, note, params)
	return result.Bundle
}

// MelodyRiff generates a melody pattern variant.
// variant "A" = base, "B" = higher density, "C" = sparse/chord-focused.
func MelodyRiff(variant string, params Params) *pflow.NetBundle {
	rng := rand.New(rand.NewSource(params.Seed))

	switch variant {
	case "B":
		// Higher density, slight seed variation
		params.Density = params.Density * 1.4
		if params.Density > 1.0 {
			params.Density = 1.0
		}
		params.Seed = params.Seed + 1000 + rng.Int63n(1000)
	case "C":
		// Sparse, longer notes
		params.Density = params.Density * 0.5
		params.Duration = int(float64(params.Duration) * 1.5)
		params.Seed = params.Seed + 2000 + rng.Int63n(1000)
	}

	result := MarkovMelody(params)
	return result.Bundle
}
